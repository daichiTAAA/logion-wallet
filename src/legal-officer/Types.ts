import { ColorTheme, rgbaToHex } from '../component/ColorTheme';
import Identity from '../component/types/Identity';
import PostalAddress from '../component/types/PostalAddress';

export const DEFAULT_LEGAL_OFFICER = "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"; // Alice
export const ANOTHER_LEGAL_OFFICER = "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty"; // Bob

const LEGAL_OFFICERS: string[] = [ DEFAULT_LEGAL_OFFICER, ANOTHER_LEGAL_OFFICER ];

export function isLegalOfficer(address: string): boolean {
    return LEGAL_OFFICERS.includes(address);
}

export type TokenizationRequestStatus = "PENDING" | "REJECTED" | "ACCEPTED";

export interface AssetDescription {
    assetId: string,
    decimals: number,
}

export interface TokenizationRequest {
    id: string,
    legalOfficerAddress: string,
    requesterAddress: string,
    requestedTokenName: string,
    bars: number,
    status: TokenizationRequestStatus,
    rejectReason?: string | null,
    createdOn?: string,
    decisionOn?: string,
    assetDescription?: AssetDescription,
}

export type LegalOfficerDecisionStatus = "PENDING" | "REJECTED" | "ACCEPTED";

export interface LegalOfficerDecision {
    legalOfficerAddress: string,
    status: LegalOfficerDecisionStatus,
    rejectReason: string | null,
    decisionOn: string | null,
}

export interface ProtectionRequest {
    id: string,
    requesterAddress: string,
    decisions: LegalOfficerDecision[],
    userIdentity: Identity,
    userPostalAddress: PostalAddress,
    createdOn: string
}

export const LIGHT_MODE: ColorTheme = {
    type: 'light',
    shadowColor: rgbaToHex('#3b6cf4', 0.1),
    dashboard: {
        background: rgbaToHex('#3b6cf4', 0.05),
        foreground: '#000000',
    },
    sidebar: {
        background: '#ffffff',
        foreground: '#000000',
        activeItemBackground: rgbaToHex('#3b6cf4', 0.15),
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#000000', 0.6),
        foreground: '#000000',
        background: rgbaToHex('#3b6cf4', 0.05),
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#ffffff',
        foreground: '#000000',
        link: rgbaToHex('#3b6cf4', 0.15),
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    }
};
