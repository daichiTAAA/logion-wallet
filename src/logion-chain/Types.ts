import BN from 'bn.js';

export type AssetId = BN;

export type AssetBalance = BN;

export interface AssetMetadata {
    name: string,
    symbol: string,
    decimals: number
}

export interface Asset {
    assetId: AssetId,
    issuer: string,
    metadata: AssetMetadata,
}

export interface AssetWithBalance {
    asset: Asset,
    balance: string,
}
