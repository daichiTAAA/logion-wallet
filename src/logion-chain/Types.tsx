export const enum ReservedName {
    LinkedLocId = "_LINKED_LOC_ID"
}

export interface MetadataItem {
    name: string;
    value: string;
}

export interface LegalOfficerCase {
    owner: string;
    requester: string;
    metadata: MetadataItem[];
    hashes: string[];
    closed: boolean;
    locType: LocType;
}

export type LocType = 'Transaction' | 'Identity';
