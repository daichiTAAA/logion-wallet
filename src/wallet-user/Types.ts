import { ColorTheme, rgbaToHex } from '../component/ColorTheme';

export const DARK_MODE: ColorTheme = {
    type: 'dark',
    shadowColor: rgbaToHex('#3b6cf4', 0.2),
    dashboard: {
        background: '#0c163d',
        foreground: '#ffffff',
    },
    sidebar: {
        background: '#152665',
        foreground: '#ffffff',
        activeItemBackground: rgbaToHex('#ffffff', 0.20),
    },
    accounts: {
        iconBackground: '#3b6cf4',
        hintColor: rgbaToHex('#ffffff', 0.6),
        foreground: '#ffffff',
        background: '#0c163d',
        legalOfficerIcon: {
            category: 'legal-officer',
            id: 'account-shield'
        }
    },
    frame: {
        background: '#152665',
        foreground: '#ffffff',
        link: '#ffffff',
    },
    topMenuItems: {
        iconGradient: {
            from: '#3b6cf4',
            to: '#6050dc',
        }
    },
    bottomMenuItems: {
        iconGradient: {
            from: '#7a90cb',
            to: '#3b6cf4',
        }
    },
    buttons: {
        secondaryBackgroundColor: '#0c163d',
    },
    select: {
        background: rgbaToHex('#ffffff', 0.20),
        foreground: '#ffffff',
        menuBackgroundColor: '#152665',
        selectedOptionBackgroundColor: rgbaToHex('#ffffff', 0.20),
    },
    table: {
        background: '#0c163d',
        foreground: '#ffffff',
        header: {
            background: '#0c163d',
            foreground: '#ffffff',
        },
        row: {
            background: '#152665',
            foreground: '#ffffff',
        }
    },
    tabs: {
        background: '#0c163d',
        foreground: '#ffffff',
        inactiveTab: {
            background: rgbaToHex('#ffffff', 0.20),
            foreground: '#ffffff',
        },
        borderColor: '#3b6cf4',
    }
};