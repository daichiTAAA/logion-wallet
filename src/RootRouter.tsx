import React from 'react';
import {
    BrowserRouter as Router,
    Redirect,
    Switch,
    Route
} from 'react-router-dom';

import { LEGAL_OFFICER_PATH, USER_PATH } from './RootPaths';
import { isLegalOfficer as isLegalOfficerFunction } from './common/types/LegalOfficer';
import LegalOfficerMain from './legal-officer/Main';
import UserMain from './wallet-user/Main';
import { useCommonContext } from './common/CommonContext';

export default function RootRouter() {
    const { currentAddress } = useCommonContext();

    if(currentAddress === "") {
        return null;
    }

    const isLegalOfficer = isLegalOfficerFunction(currentAddress);
    let redirectTo;
    if(isLegalOfficer) {
        redirectTo = LEGAL_OFFICER_PATH;
    } else {
        redirectTo = USER_PATH;
    }

    return (
        <Router>
            <Switch>
                <Route path={ LEGAL_OFFICER_PATH }>
                    { isLegalOfficer ? <LegalOfficerMain /> : <Redirect to={ USER_PATH } /> }
                </Route>
                <Route path={ USER_PATH }>
                    { !isLegalOfficer ? <UserMain /> : <Redirect to={ LEGAL_OFFICER_PATH } /> }
                </Route>
                <Route path="/">
                    <Redirect to={ redirectTo } />
                </Route>
            </Switch>
        </Router>
    );
}
