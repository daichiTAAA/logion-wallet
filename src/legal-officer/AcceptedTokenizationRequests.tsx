import React from 'react';

import Table from 'react-bootstrap/Table';

import { useLegalOfficerContext } from './LegalOfficerContext';

export default function AcceptedTokenizationRequests() {
    const { acceptedTokenizationRequests } = useLegalOfficerContext();

    if(acceptedTokenizationRequests === null) {
        return null;
    }

    return (
        <>
            <h3>Accepted</h3>
            <Table striped bordered>
                <thead>
                    <tr>
                        <th>Requester</th>
                        <th>Token</th>
                        <th>Bars</th>
                        <th>Timestamp</th>
                    </tr>
                </thead>
                <tbody>
                    {
                        acceptedTokenizationRequests.map(request => (
                            <tr key={request.id}>
                                <td>{request.requesterAddress}</td>
                                <td>{request.requestedTokenName}</td>
                                <td>{request.bars}</td>
                                <td>{request.decisionOn || ""}</td>
                            </tr>
                        ))
                    }
                </tbody>
            </Table>
        </>
    );
}
