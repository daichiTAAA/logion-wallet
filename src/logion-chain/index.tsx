// Imports
import { KeyringPair as PolkadotKeyringPair } from '@polkadot/keyring/types';
import {
    Block as PolkadotBlock,
    Hash as PolkadotHash,
    Header as PolkadotHeader,
    SignedBlock as PolkadotSignedBlock,
    EventRecord as PolkadotEventRecord,
    Extrinsic as PolkadotExtrinsic,
} from '@polkadot/types/interfaces';
import keyring from '@polkadot/ui-keyring';
import { InjectedAccountWithMeta as PolkadotInjectedAccountWithMeta } from '@polkadot/extension-inject/types';
import { ISubmittableResult as PolkadotISubmittableResult } from '@polkadot/types/types';

import {
    LogionChainContextProvider,
    useLogionChain,

    ExtrinsicFetchSpecification as ContextExtrinsicFetchSpecification,
    ApiState as ContextApiState,
    NodeMetadata as ContextNodeMetadata,
} from './LogionChainContext';

import {
    signAndSend,
    replaceUnsubscriber,
    sign,

    ExtrinsicSignatureParameters as SignatureExtrinsicSignatureParameters,
    Unsubscriber as SignatureUnsubscriber,
    SignAndSendCallback as SignatureSignAndSendCallback,
    AttributesSignatureParameters as SignatureStringSignatureParameters,
} from './Signature';

import {
    getTimestamp,
    fetchExtrinsics,
    hashEquals,

    ExtrinsicsAndHead as BlocksExtrinsicsAndHead,
} from './Blocks';

import {
    isExtensionAvailable,
    recommendedExtension
} from './Keys';

// Re-exports
export {
    useLogionChain,
    LogionChainContextProvider,
    getTimestamp,
    fetchExtrinsics,
    hashEquals,
    signAndSend,
    replaceUnsubscriber,
    isExtensionAvailable,
    recommendedExtension,
    keyring,
    sign,
};

export type Unsubscriber = SignatureUnsubscriber;
export type ExtrinsicSignatureParameters = SignatureExtrinsicSignatureParameters;
export type ExtrinsicFetchSpecification = ContextExtrinsicFetchSpecification;

export type ExtrinsicsAndHead = BlocksExtrinsicsAndHead;

export type KeyringPair = PolkadotKeyringPair;
export type Hash = PolkadotHash;
export type SignedBlock = PolkadotSignedBlock;
export type Header = PolkadotHeader;
export type Block = PolkadotBlock;
export type EventRecord = PolkadotEventRecord;
export type Extrinsic = PolkadotExtrinsic;
export type SignAndSendCallback = SignatureSignAndSendCallback;
export type ApiState = ContextApiState;
export type NodeMetadata = ContextNodeMetadata;
export type InjectedAccountWithMeta = PolkadotInjectedAccountWithMeta;
export type StringSignatureParameters = SignatureStringSignatureParameters;
export type ISubmittableResult = PolkadotISubmittableResult;
