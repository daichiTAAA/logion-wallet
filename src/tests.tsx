import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
