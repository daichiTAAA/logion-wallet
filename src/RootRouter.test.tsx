jest.mock('./logion-chain');
jest.mock('./RootContext');

import { setContextMock } from './logion-chain';
import { shallowRender, mockAccount, act } from './tests';

import RootRouter from './RootRouter';
import { DEFAULT_LEGAL_OFFICER } from './legal-officer/Types';

test('Given null injected accounts, when rendering, then null', () => {
    setContextMock({
        injectedAccounts: null,
    });
    let tree: any;
    act(() => { tree = shallowRender(<RootRouter />) });
    expect(tree).toMatchSnapshot();
});

test('Given no injected accounts, when rendering, then null', () => {
    setContextMock({
        injectedAccounts: [],
    });
    let tree: any;
    act(() => { tree = shallowRender(<RootRouter />) });
    expect(tree).toMatchSnapshot();
});

test('Given legal officer, when rendering, then route to wallet or redirect', () => {
    setContextMock({
        injectedAccounts: [
            mockAccount(DEFAULT_LEGAL_OFFICER, "Account name")
        ]
    });

    let tree: any;
    act(() => { tree = shallowRender(<RootRouter />) });

    expect(tree).toMatchSnapshot();
});

test('Given wallet user, when rendering, then route to wallet or redirect', () => {
    setContextMock({
        injectedAccounts: [
            mockAccount("userAddress", "Account name")
        ]
    });

    let tree: any;
    act(() => { tree = shallowRender(<RootRouter />) });

    expect(tree).toMatchSnapshot();
});
