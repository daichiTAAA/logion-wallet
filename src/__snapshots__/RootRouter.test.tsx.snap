// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given legal officer, when rendering, then render legal officer UI or redirect to login 1`] = `
<Routes>
  <Route
    element={
      <RenderOrRedirectToLogin
        render={[Function]}
      />
    }
    path="/legal-officer/*"
  />
  <Route
    element={
      <Navigate
        to="/legal-officer"
      />
    }
    path="/user/*"
  />
  <Route
    element={<Login />}
    path="/login"
  />
  <Route
    element={
      <Navigate
        to="/legal-officer"
      />
    }
    path="/"
  />
</Routes>
`;

exports[`Given no addresses, when rendering, then null 1`] = `null`;

exports[`Given null addresses, when rendering, then null 1`] = `null`;

exports[`Given undefined current address, when rendering, then enable redirect to login via either legal officer or user route 1`] = `
<Routes>
  <Route
    element={
      <Navigate
        to="/user"
      />
    }
    path="/legal-officer/*"
  />
  <Route
    element={
      <RenderOrRedirectToLogin
        render={[Function]}
      />
    }
    path="/user/*"
  />
  <Route
    element={<Login />}
    path="/login"
  />
  <Route
    element={
      <Navigate
        to="/user"
      />
    }
    path="/"
  />
</Routes>
`;

exports[`Given wallet user, when rendering, then render user UI or redirect to login 1`] = `
<Routes>
  <Route
    element={
      <Navigate
        to="/user"
      />
    }
    path="/legal-officer/*"
  />
  <Route
    element={
      <RenderOrRedirectToLogin
        render={[Function]}
      />
    }
    path="/user/*"
  />
  <Route
    element={<Login />}
    path="/login"
  />
  <Route
    element={
      <Navigate
        to="/user"
      />
    }
    path="/"
  />
</Routes>
`;
